xonsh/events.py

```diff
@@ -0,0 +1,101 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.doc("Comes with eggs")
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guaranteed.
+    """
+
+    @classmethod
+    def doc(cls, text):
+        cls.__doc__ = text
+
+    def __call__(self, func):
```

Comment:
this `__call__` trick is a little too implicit imo, we can have a decorator that does the same thing with just a module level decorator function.